### PR TITLE
Custom sample filtering performance and security improvement

### PR DESCRIPTION
**Pull request by dippindots**
Fix https://github.com/cBioPortal/cbioportal/issues/11195

- `concat` has poor performance during my test, we should try to not use `concat` in a large `foreach` clause
- clickhouse support native `array` in the query, and it has better performance, try replace large foreach with `array`

**In this pull request:**
- Change back from string to prepared statement for security reason
- Use native `array` to improve performance

### CODE REVIEW

File: src/main/java/org/cbioportal/persistence/helper/StudyViewFilterHelper.java [hunk 1]
```diff
@@ -65,6 +66,13 @@ private StudyViewFilterHelper(@NonNull StudyViewFilter studyViewFilter,
         this.categorizedGenericAssayDataCountFilter = extractGenericAssayDataCountFilters(studyViewFilter, genericAssayProfilesMap);
         this.customDataSamples = customDataSamples;
         this.involvedCancerStudies = involvedCancerStudies;
+        if (studyViewFilter != null && studyViewFilter.getSampleIdentifiers() != null) {
```
[the remark this answers is not in the document]
Comment (haynescd) [reply src/main/java/org/cbioportal/persistence/helper/StudyViewFilterHelper.java:69]: Yea, so Java just recently implemented this concept of Property you are talking about. With Records.

But I agree this shouldn't be in the constructor and we could just add this to the filteredSampleIdentifiers method to dynamically calculate this

File: src/main/java/org/cbioportal/web/parameter/CustomSampleIdentifier.java [hunk 2]
```diff
@@ -22,4 +24,14 @@ public String getValue() {
...
     }
+
+    // Generating unique SampleId by concatenating studyId and sampleId
+    public String getUniqueSampleId() {
+        // Assuming studyId and sampleId are available in SampleIdentifier
+        // Concatenate with "_" in between if both values are not null
+        if (getStudyId() != null && getSampleId() != null) {
+            return getStudyId() + "_" + getSampleId();
+        }
+        return null;  // or return a default value if either studyId or sampleId is null
```
Comment (haynescd) [inline src/main/java/org/cbioportal/web/parameter/CustomSampleIdentifier.java:35]: I would just update the comment to reflect the default value is null

File: src/main/java/org/cbioportal/web/parameter/CustomSampleIdentifier.java [hunk 3]
```diff
@@ -22,4 +24,14 @@ public String getValue() {
     public void setValue(String value) {
         this.value = value;
     }
+
+    // Generating unique SampleId by concatenating studyId and sampleId
+    public String getUniqueSampleId() {
+        // Assuming studyId and sampleId are available in SampleIdentifier
+        // Concatenate with "_" in between if both values are not null
+        if (getStudyId() != null && getSampleId() != null) {
```
Comment (haynescd) [inline src/main/java/org/cbioportal/web/parameter/CustomSampleIdentifier.java:32]: Can this even happen? a sample Id with one of these null?

File: src/main/java/org/cbioportal/persistence/helper/StudyViewFilterHelper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (alisman) [reply src/main/java/org/cbioportal/persistence/helper/StudyViewFilterHelper.java:69]: But if it's called multiple times then, absent some kind of mobx-like computed thing, we should just leave it.  I think that's the existing pattern with "build", right?  You do the calculation up front?

**Pull request merged**